lib/src/main/java/io/ably/lib/rest/Auth.java

```diff
@@ -827,13 +829,30 @@ public AuthOptions getAuthOptions() {
      * Renew auth credentials.
      * Will obtain a new token, even if we already have an apparently valid one.
      * Authorization will use the parameters supplied on construction.
+     * @deprecated Use {@link Auth#renewAuth} instead
      */
+    @Deprecated
     public TokenDetails renew() throws AblyException {
         TokenDetails tokenDetails = assertValidToken(this.tokenParams, this.authOptions, true);
         ably.onAuthUpdated(tokenDetails.token, false);
         return tokenDetails;
     }
 
+    /**
+     * Renew auth credentials.
+     * Will obtain a new token, even if we already have an apparently valid one.
+     * Authorization will use the parameters supplied on construction.
+     *
+     * @return
+     * A single entry that contain a token detail and a future that represent an asynchronous result
+     * Clients must wait for the future result to finish before processing. If there is an exception happened during
+     * asynchronous operation the future will contain an AblyException
+     */
+    public Map.Entry<TokenDetails, Future<AblyException>> renewAuth() throws AblyException {
```

Comment:
This interface definition looks wrong to me, in that there's an [in-built exception handling path](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#completeExceptionally-java.lang.Throwable-) provided by `Future`s, therefore it's not idiomatic to return an exception as the result.

It's been a while since I've stared at `Future`-based APIs, or tried to define one myself, but I think the standard return type when you've got nothing to return except for the notion of 'success' is `CompletableFuture<Void>`... [this Stackoverflow answer](https://stackoverflow.com/a/39002692/392847) seems to capture that quite well.

Reply:
Fair enough, I did not think it was ideal at all. I have actually initially used a Future<Void> than decided that it would be pragmatic to return an optional exception. I will see if I can use CompletableFuture and provide a more appropriate error path.

Reply:
Unfortunately we cannot yet use [CompletableFuture](https://developer.android.com/reference/java/util/concurrent/CompletableFuture) as it's only available on Android SDK >=24 
However I have changed the API surface a little bit and threw exception directly from `call()` . According to this [SO answer ](https://stackoverflow.com/a/2248166) we are permitted to throw checked exception from call that is going to be propagated as an `ExecutionException` to the caller.  Change : 14359cea0406dddb895873c8d0a3e66ab6fb9357